### ISSUE TITLE
Search not working as expected with screen readers

### ISSUE THREAD

**Issue by nickcolley** [alphagov/govuk-design-system#897]
This issue is from a May 2019 external accessibility audit report.

**WCAG Reference**:
- [1.3.1 Info and Relationships (Level A)](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html)
- [4.1.2 Name, Role, Value (Level A)](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value.html)
**Issue ID**: DAC_4
**URL**: All pages

## Summary
The autocomplete search feature caused multiple issues for screen reader users.

There were many issues encountered with the search feature that affected screen reader users in different ways depending on the software and browser used.

## Screen Shot
![](https://user-images.githubusercontent.com/2445413/58793362-d80d3b80-85ed-11e9-88ec-caac8d465823.png)

Many issues were encountered by screen reader users with regards to the search feature.
These issues varied depending on which screen reader and browser were used and caused great difficulties for screen reader users but no other users of assistive technologies.

## Solution
Ensure that the search feature works as expected, allowing all users to perform a search of the website. The following autocomplete is the most accessible feature currently tested and although there are some issues with Internet Explorer, we believe that these are being addressed.
URL: https://haltersweb.github.io/Accessibility/autocomplete.html

## Related stories
- [ ] Search button disappearing can be confusing for some users #937
- [ ] Beta banner is confusing when placed in between menu button and menu items #939
- [ ] Navigating to search via the tab key in JAWS results in unexpected behaviour #938
- [ ] Searches without relevant results can return results #940
- [ ] JAWS and NVDA form fields dialogue do not show autocomplete clearly [#344 (accessible-autocomplete)](https://github.com/alphagov/accessible-autocomplete/issues/344)
- [ ] JAWS 17/18 doesn't announce the number of results available [#235 (accessible-autocomplete)](https://github.com/alphagov/accessible-autocomplete/issues/235)
- [ ] Overlayed options cannot be swiped to on VoiceOver iOS [#345 (accessible-autocomplete)](https://github.com/alphagov/accessible-autocomplete/issues/345)
- [ ] NVDA unable to use 'escape' to exit the autocomplete [#346 (accessible-autocomplete)](https://github.com/alphagov/accessible-autocomplete/issues/346)

**Comment by kellylee-gds** [comment]
This needs further investigation to look at whether the issues are due to the accessible autocomplete or with the search implementation. We should also create new stories of issues found and/or identify existing issues in the autocomplete repo.

**Comment by nickcolley** [comment]
1. Try to reproduce the issues reported in the design system website.
2. Try to reproduce the issues reported in the vanilla stock accessible-autocomplete project.
3. Work out the simplest test cases.
4. File bugs in the relevant place ready for triage.

**Comment by nickcolley** [comment]
In regards to the cognitive user comments, I think we've had some feedback before that the icon is a bit confusing in the past...

We could consider making it a bit like DuckDuckGo: https://duckduckgo.com/?q=&t=hd

<img width="613" alt="Screen Shot 2019-06-18 at 17 55 27" src="https://user-images.githubusercontent.com/2445413/59703661-46f3b280-91f2-11e9-8e9e-d54317ebf47f.png">

When the input is not being used there's a search icon, if clicked on it'll focus the overall input.
<img width="318" alt="Screen Shot 2019-06-18 at 17 49 38" src="https://user-images.githubusercontent.com/2445413/59703367-9f768000-91f1-11e9-9e1a-11509498303b.png">

Different to DuckDuckGo the icon would not turn into a button, and just stick around.

<img width="339" alt="Screen Shot 2019-06-18 at 17 55 13" src="https://user-images.githubusercontent.com/2445413/59703649-3d6a4a80-91f2-11e9-9ed6-27157de0d8bc.png">

**Comment by nickcolley** [comment]
Regarding the feedback from 'Navigating via the tab key' using JAWs.

> When using JAWS, I navigated with the ‘Tab’ key and located a link which announced as ‘Search design system’. I selected the link and was taken directly to an autocomplete to search for a component.

I've managed to reproduce this issue in Internet Explorer 11 and using JAWS 2018, it seems to be a bug with JAWS where the link's announcement is duplicated onto the input's announcement.

Tab to Design System home link
Announcement: "GOV.UK Design System link"
Tab to 'Search design system' input
Announcement: "Search design system visited link"
Shift tab to Design System home link
Announcement: "GOV.UK Design System link"
Tab to 'Search design system' input
Announcement: "Search design system edit combo [...]"

TODO: PUT VIDEO HERE

This does not happen when tabbing away from the input and tabbing back towards in the input.

Minimal markup to reproduce the issue:

```html
    <a href="/">Home</a>
    <label for="username">Username</label>
    <div role="combobox">
      <input id="username">
    </div>
```

When removing the `[role=combobox]` wrapper, the problem no longer exists.

**Comment by nickcolley** [comment]
I'm going to start breaking this issue up into smaller issues, then converting this into an epic style card.

**Comment by nickcolley** [comment]
I'm not sure what to do with the following feedback as I don't understand what 'content' and 'arrow keys' mean together here:

> “I attempted to view the label of the autocomplete while situated both in and out of context using JAWS and found that the content which announced did not provide a clear indication of the purpose of the autocomplete. While in context with the arrow keys I found that multiple items were located which appeared to be marked up as edit combos and announced either as ‘No search results’ Or ‘Search design system’.

It is possible that other improvements to ensuring the input being put into the form fields list correctly would resolve these issues as they seem related.

This is something to contact the external auditor for more information on.

**Comment by nickcolley** [comment]
I have moved all actionable things into 'related stories' as an epic, there are a few bits of feedback I have not created cards for because they seemed related to the others or too vague to action.

My suggestion would be to focus on the actionable work, and try and identify any improvements while we do that (HMRC's work seems great in this respect), then if we have budget to do a re-test of the search when we're more confident with the work we've done.

**Comment by 36degrees** [comment]
Thanks @nickcolley for breaking this down into individual issues 👍 

I've collated them together under a new epic #965 which we'll be aiming to pick up this quarter.

I think this can now be closed.

**Comment by nickcolley** [comment]
Confirmed as fixed in a re-test by DAC on 7th October 2019
